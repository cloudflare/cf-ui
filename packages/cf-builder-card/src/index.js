import cardActions from './cardActions';
import CardActionTypes from './CardActionTypes';
import CardBuilder from './CardBuilder';
import cardReducer from './cardReducer';

export { cardActions, CardActionTypes, CardBuilder, cardReducer };
