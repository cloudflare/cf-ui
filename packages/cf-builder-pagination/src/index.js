import PaginationBuilder from './PaginationBuilder';

export { PaginationBuilder };
