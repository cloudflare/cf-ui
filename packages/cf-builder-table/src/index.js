import tableActions from './tableActions';
import TableActionTypes from './TableActionTypes';
import TableBuilder from './TableBuilder';
import TableBuilderPropTypes from './TableBuilderPropTypes';
import tableReducer from './tableReducer';

export {
  tableActions,
  TableActionTypes,
  TableBuilder,
  TableBuilderPropTypes,
  tableReducer
};
