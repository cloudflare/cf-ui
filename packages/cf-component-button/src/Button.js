import React, { PropTypes } from 'react';
import { createComponent } from 'cf-style-container';

const getRadius = (borderRadius, group) => {
  if (group) {
    return {
      borderTopRightRadius: group === 'last' ? borderRadius : 0,
      borderTopLeftRadius: group === 'first' ? borderRadius : 0,
      borderBottomLeftRadius: group === 'first' ? borderRadius : 0,
      borderBottomRightRadius: group === 'last' ? borderRadius : 0
    };
  }
  return { borderRadius };
};

const getOpacity = props => {
  if (props.loading) {
    return 0.8;
  }
  if (props.disabled) {
    return 0.5;
  }
  return 'inherit';
};

const getLoadingBefore = (fadeZoomIn, loading) => {
  if (!loading) {
    return null;
  }
  // this needs something better (e.g. a rotating icon)
  return {
    '&::before': {
      content: '"â€¦"',
      color: 'black',
      display: 'block',
      position: 'absolute',
      top: '0.6em',
      left: 0,
      height: '100%',
      width: '100%',
      opacity: 1
    }
  };
};

const styles = props => {
  const theme = props.theme;
  return Object.assign(
    {
      '&:hover': {
        backgroundColor: props.loading
          ? theme.disabledBackground
          : theme[`${props.type}HoverBackground`],
        borderColor: theme[`${props.type}HoverBorder`],
        color: props.loading
          ? theme.disabledBackground
          : theme[`${props.type}HoverColor`]
      },
      '&:active': {
        backgroundColor: theme[`${props.type}ActiveBackground`],
        borderColor: theme[`${props.type}ActiveBorder`],
        color: theme[`${props.type}ActiveColor`]
      },
      '&:focus': {
        backgroundColor: props.loading
          ? theme.backgroundColor
          : theme[`${props.type}Background`],
        boxShadow: `inset 0px 0px 0px ${theme.borderSize} ${theme[`${props.type}focusOutlineColor`]}`,
        color: props.loading
          ? theme.disabledBackground
          : theme[`${props.type}FocusColor`],
        outline: props.loading ? 'none' : 'inherit'
      },
      '&[title]': {
        pointerEvents: props.disabled ? 'auto' : 'none'
      },
      backgroundColor: props.loading
        ? theme.disabledBackground
        : theme[`${props.type}Background`],
      borderBottom: theme.borderBottom,
      borderBottomWidth: theme.borderBottomWidth,
      borderColor: theme[`${props.type}Border`],
      borderLeft: props.group && props.group !== 'first' ? 0 : theme.borderLeft,
      borderRight: theme.borderRight,
      borderStyle: theme.borderStyle,
      borderTop: theme.borderTop,
      borderWidth: theme.borderWidth,
      color: props.loading
        ? theme.disabledBackground
        : theme[`${props.type}Color`],
      cursor: props.disabled || props.loading ? 'default' : theme.cursor,
      display: theme.display,
      fontFamily: theme.fontFamily,
      fontSize: theme.fontSize,
      fontWeight: theme.fontWeight,
      lineHeight: theme.lineHeight,
      marginBottom: props.group ? 0 : theme.marginBottom,
      marginLeft: props.group ? 0 : theme.marginLeft,
      marginRight: props.group ? 0 : theme.marginRight,
      marginTop: props.group ? 0 : theme.marginTop,
      opacity: getOpacity(props),
      paddingBottom: theme.paddingBottom,
      paddingLeft: theme.paddingLeft,
      paddingRight: theme.paddingRight,
      paddingTop: theme.paddingTop,
      pointerEvents: props.disabled ? 'none' : 'auto',
      position: theme.position,
      textAlign: theme.textAlign,
      transition: theme.transition,
      userSelect: theme.userSelect
    },
    getRadius(theme.borderRadius, props.group),
    getLoadingBefore(theme.fadeZoomIn, props.loading)
  );
};

const Button = (
  {
    onClick,
    submit,
    className,
    disabled,
    loading,
    children
  }
) => (
  <button
    type={submit ? 'submit' : 'button'}
    disabled={disabled || loading}
    className={className}
    onClick={onClick}
  >
    {children}
  </button>
);

Button.propTypes = {
  onClick: PropTypes.func.isRequired,
  submit: PropTypes.bool,
  className: PropTypes.string.isRequired,
  group: PropTypes.oneOf(['first', 'inbetween', 'last']),
  type: PropTypes.oneOf([
    'default',
    'primary',
    'success',
    'warning',
    'danger'
  ]).isRequired,
  disabled: PropTypes.bool,
  loading: PropTypes.bool,
  children: PropTypes.node
};

Button.defaultProps = {
  type: 'default',
  submit: false
};

export default createComponent(styles, Button);
