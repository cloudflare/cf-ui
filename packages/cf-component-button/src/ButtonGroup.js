import React, {PropTypes} from 'react';
import {createComponent} from 'cf-style-container';

const styles = props => {
  const theme = props.theme;
  return ({
    display: theme.display,
    position: theme.position
  });
};

const getGroupByIndex = (index, length) => {
  if (index === length - 1) {
    return 'last';
  }
  if (index === 0) {
    return 'first';
  }
  return 'inbetween';
};

const addGroupProps = children =>
  React.Children.map(children, (child, index) =>
    React.cloneElement(child, {
      group: getGroupByIndex(index, React.Children.count(children))
    })
  );

const ButtonGroup = ({children, className}) =>
  <div className={className}>{addGroupProps(children)}</div>;

ButtonGroup.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string.isRequired
};

export default createComponent(styles, ButtonGroup);
