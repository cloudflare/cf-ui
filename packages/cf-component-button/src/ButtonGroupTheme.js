export default {
  display: 'inline-block',
  position: 'relative'
};
