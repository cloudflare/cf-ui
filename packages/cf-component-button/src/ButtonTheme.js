import { darken, lighten } from 'polished';

export default baseTheme => {
  const defaultBorder = darken(
    baseTheme.colorOffsetDark,
    baseTheme.colorGrayLight
  );
  const primaryBorder = darken(baseTheme.colorOffsetDark, baseTheme.colorBlue);
  const successBorder = darken(baseTheme.colorOffsetDark, baseTheme.colorGreen);
  const warningBorder = darken(
    baseTheme.colorOffsetDark,
    baseTheme.colorOrange
  );
  const dangerBorder = darken(baseTheme.colorOffsetDark, baseTheme.colorRed);
  const defaultHoverBackground = darken(
    baseTheme.colorOffsetDark,
    baseTheme.colorGrayLight
  );
  const defaultActiveBackground = darken(
    baseTheme.colorOffsetDark,
    defaultHoverBackground
  );
  const primaryActiveBackground = darken(
    baseTheme.colorOffsetDark,
    baseTheme.colorBlueDark
  );
  const successActiveBackground = darken(
    baseTheme.colorOffsetDark,
    baseTheme.colorGreenDark
  );
  const warningActiveBackground = darken(
    baseTheme.colorOffsetDark,
    baseTheme.colorOrangeDark
  );
  const dangerActiveBackground = darken(
    baseTheme.colorOffsetDark,
    baseTheme.colorRedDark
  );

  return {
    fontSize: `${(parseFloat(baseTheme.fontSize) - 1) / parseFloat(baseTheme.fontSize)}rem`,
    fontWeight: baseTheme.weightNormal,
    paddingTop: '0.6em',
    paddingRight: '1em',
    paddingBottom: '0.53em',
    paddingLeft: '1em',
    spacing: '0.4rem',
    borderSize: '1px',
    display: 'inline-block',
    position: 'relative',
    marginTop: 0,
    marginBottom: 0,
    marginLeft: 0,
    marginRight: 0,
    background: 'transparent',
    lineHeight: 1.2,
    textAlign: 'center',
    webkitTextStroke: 0, // This makes white text in buttons hard to read, so removing it
    webkitFontSmoothing: 'subpixel-antialiased',
    userSelect: 'none',
    cursor: 'pointer',
    transition: 'all 200ms ease',
    border: 'none',
    borderLeft: 0,
    borderBottom: '1px solid',
    borderTop: 0,
    borderRight: 0,

    // Disabled
    disabledBorder: darken(
      baseTheme.colorOffsetDark,
      baseTheme.disabledBackground
    ),
    disabledColor: lighten(0.3, baseTheme.disabledBackground),

    // Normal
    defaultBackground: baseTheme.colorGrayLight,
    primaryBackground: baseTheme.colorBlue,
    successBackground: baseTheme.colorGreen,
    warningBackground: baseTheme.colorOrange,
    dangerBackground: baseTheme.colorRed,

    defaultBorder,
    primaryBorder,
    successBorder,
    warningBorder,
    dangerBorder,

    defaultColor: baseTheme.colorGrayDark,
    primaryColor: baseTheme.colorWhite,
    successColor: baseTheme.colorWhite,
    warningColor: baseTheme.colorWhite,
    dangerColor: baseTheme.colorWhite,

    // Hover
    defaultHoverBackground,
    primaryHoverBackground: baseTheme.colorBlueDark,
    successHoverBackground: baseTheme.colorGreenDark,
    warningHoverBackground: baseTheme.colorOrangeDark,
    dangerHoverBackground: baseTheme.colorRedDark,

    defaultHoverBorder: darken(baseTheme.colorOffsetDark, defaultBorder),
    primaryHoverBorder: darken(baseTheme.colorOffsetDark, primaryBorder),
    successHoverBorder: darken(
      baseTheme.colorOffsetDark,
      baseTheme.colorGreenDark
    ),
    warningHoverBorder: darken(
      baseTheme.colorOffsetDark,
      baseTheme.colorOrangeDark
    ),
    dangerHoverBorder: darken(
      baseTheme.colorOffsetDark,
      baseTheme.colorRedDark
    ),

    defaultHoverColor: baseTheme.colorGrayDark,
    primaryHoverColor: baseTheme.colorWhite,
    successHoverColor: baseTheme.colorWhite,
    warningHoverColor: baseTheme.colorWhite,
    dangerHoverColor: baseTheme.colorWhite,

    // Active
    defaultActiveBackground,
    primaryActiveBackground,
    successActiveBackground,
    warningActiveBackground,
    dangerActiveBackground,

    defaultActiveBorder: darken(
      baseTheme.colorOffsetDark,
      defaultActiveBackground
    ),
    primaryActiveBorder: darken(
      baseTheme.colorOffsetDark,
      primaryActiveBackground
    ),
    successActiveBorder: darken(
      baseTheme.colorOffsetDark,
      successActiveBackground
    ),
    warningActiveBorder: darken(
      baseTheme.colorOffsetDark,
      warningActiveBackground
    ),
    dangerActiveBorder: darken(
      baseTheme.colorOffsetDark,
      dangerActiveBackground
    ),

    defaultActiveColor: baseTheme.colorGrayDark,
    primaryActiveColor: baseTheme.colorWhite,
    successActiveColor: baseTheme.colorWhite,
    warningActiveColor: baseTheme.colorWhite,
    dangerActiveColor: baseTheme.colorWhite,

    // Focus
    defaultFocusColor: baseTheme.colorGrayDark,
    primaryFocusColor: baseTheme.colorWhite,
    successFocusColor: baseTheme.colorWhite,
    warningFocusColor: baseTheme.colorWhite,
    dangerFocusColor: baseTheme.colorWhite,

    defaultFocusOutlineColor: darken(
      baseTheme.colorOffsetDark,
      defaultActiveBackground
    ),
    primaryFocusOutlineColor: darken(
      baseTheme.colorOffsetDark,
      primaryActiveBackground
    ),
    successFocusOutlineColor: darken(
      baseTheme.colorOffsetDark,
      successActiveBackground
    ),
    warningFocusOutlineColor: darken(
      baseTheme.colorOffsetDark,
      warningActiveBackground
    ),
    dangerFocusOutlineColor: darken(
      baseTheme.colorOffsetDark,
      dangerActiveBackground
    )
  };
};
