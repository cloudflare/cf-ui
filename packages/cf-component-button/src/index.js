import Button from './Button';
import ButtonTheme from './ButtonTheme';
import ButtonGroup from './ButtonGroup';
import ButtonGroupTheme from './ButtonGroupTheme';

export { Button, ButtonTheme, ButtonGroup, ButtonGroupTheme };
