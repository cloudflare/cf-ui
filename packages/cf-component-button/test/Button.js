import React from 'react';
import { felaSnapshot } from 'cf-style-provider';
import { Button } from '../../cf-component-button/src/index';

test('should render with type', () => {
  const snapshot = felaSnapshot(
    <Button type="primary" onClick={() => {}}>Button</Button>
  );
  expect(snapshot.component).toMatchSnapshot();
  expect(snapshot.styles).toMatchSnapshot();
});

test('should render with type/disabled', () => {
  const snapshot = felaSnapshot(
    <Button type="primary" disabled onClick={() => {}}>Button</Button>
  );
  expect(snapshot.component).toMatchSnapshot();
  expect(snapshot.styles).toMatchSnapshot();
});

test('should render as submit', () => {
  const snapshot = felaSnapshot(
    <Button type="primary" submit onClick={() => {}}>Button</Button>
  );
  expect(snapshot.component).toMatchSnapshot();
  expect(snapshot.styles).toMatchSnapshot();
});

test('should render with loading', () => {
  const snapshot = felaSnapshot(
    <Button type="primary" loading onClick={() => {}}>Button</Button>
  );
  expect(snapshot.component).toMatchSnapshot();
  expect(snapshot.styles).toMatchSnapshot();
});

test('should render with loading and disabled overridden', () => {
  const snapshot = felaSnapshot(
    <Button type="primary" loading disabled={false} onClick={() => {}}>
      Button
    </Button>
  );
  expect(snapshot.component).toMatchSnapshot();
  expect(snapshot.styles).toMatchSnapshot();
});
