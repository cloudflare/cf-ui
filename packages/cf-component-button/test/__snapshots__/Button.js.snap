// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render as submit 1`] = `
<button
  className="cf-yp66z7"
  disabled={undefined}
  onClick={[Function]}
  type="submit"
>
  Button
</button>
`;

exports[`should render with loading 1`] = `
<button
  className="cf-uydyf0"
  disabled={true}
  onClick={[Function]}
  type="button"
>
  Button
</button>
`;

exports[`should render with loading and disabled overridden 1`] = `
<button
  className="cf-uydyf0"
  disabled={true}
  onClick={[Function]}
  type="button"
>
  Button
</button>
`;

exports[`should render with type 1`] = `
<button
  className="cf-yp66z7"
  disabled={undefined}
  onClick={[Function]}
  type="button"
>
  Button
</button>
`;

exports[`should render with type/disabled 1`] = `
<button
  className="cf-1903je5"
  disabled={true}
  onClick={[Function]}
  type="button"
>
  Button
</button>
`;
